src/devices/src/virtio/balloon/utils.rs

```diff
@@ -25,25 +26,33 @@ pub(crate) fn compact_page_frame_numbers(v: &mut [u32]) -> Vec<(u32, u32)> {
     let mut result = Vec::with_capacity(MAX_PAGE_COMPACT_BUFFER);
 
     // The most recent range of pages is [previous..previous + length).
-    let mut previous = v[0];
+    let mut previous = 0;
     let mut length = 1;
 
-    for page_frame_number in &v[1..] {
+    for pfn_index in 1..v.len() {
+        let page_frame_number = v[pfn_index];
+
+        // Skip duplicate pages.
+        if page_frame_number == v[pfn_index - 1] {
+            error!("Skipping duplicate PFN.");
```

Comment:
You can also print out the duplicate PFN.

Reply:
Done.